### PR TITLE
Upgrade to the Firebase 3.2.0 SDK

### PR DESCRIPTION
**Pull request by HeyDaniel**
now you can use Firebase3.2.0 now

### CODE REVIEW

**Comment by ksinghal** [comment]
You may want to update the Podspec dependency to 3.2.0 along with these changes since it currently depends on 2.2. Just a heads up thanks for the good work I am excited to upgrade once GeoFire is compatible!

**Comment by zgosalvez** [comment]
Tagging issue #39

**Comment by greggmojica** [comment]
@mcdonamp Any idea when this will be merged?

**Comment by jdimond** [comment]
@HeyDaniel needs to sign the CLA first before we can merge

File: GeoFire/GFCircleQuery.h [hunk 1]
```diff
@@ -20,4 +20,9 @@
  */
 @property (atomic, readwrite) double radius;
 
+//private method
+- (id)initWithGeoFire:(GeoFire *)geoFire
+             location:(CLLocation *)location
+               radius:(double)radius;
+
```
Comment (jdimond) [inline GeoFire/GFCircleQuery.h]: Any reason you chose to get rid of the private headers? This allows a clear separation of API and internal methods and makes it easier to document and understand what the API surface is.

**Comment by mcdonamp** [comment]
Let's revert all the changes like `GeoFire/Implementation/GFBase32Utils.h → GeoFire/GFBase32Utils.h` as well--as @jdimond mentioned, we like to keep the implementation separate from the API, partially for our own sanity, but also because we've got other tools that generate public headers, etc. that are relying on these differences.

File: GeoFire/GeoFire.h [hunk 2]
```diff
@@ -128,4 +128,11 @@ withCompletionBlock:(GFCompletionBlock)block;
  */
 - (GFRegionQuery *)queryWithRegion:(MKCoordinateRegion)region;
 
+
+// private method
+- (FIRDatabaseReference *)firebaseRefForLocationKey:(NSString *)key;
+
++ (CLLocation *)locationFromValue:(id)dict;
+// 这个方法好像没有实现+ (NSDictionary *)dictFromLocation:(CLLocation *)location;
```
Comment (mcdonamp) [inline GeoFire/GeoFire.h]: Clean up the comment?

**Comment by mcdonamp** [comment]
As mentioned above, we'll also want to change the Podspec:

```ruby
Pod::Spec.new do |s|
  s.name         = "GeoFire"
  s.version      = "2.0.0"
  s.summary      = "Realtime location queries with Firebase."
  s.homepage     = "https://github.com/firebase/geofire-objc"
  s.license      = { :type => 'MIT', :file => 'LICENSE' }
  s.author       = { "Firebase" => "https://firebase.google.com/support" }
  s.source       = { :git => "https://github.com/firebase/geofire-objc.git", :tag => 'v2.0.0' }
  s.source_files = "GeoFire/**/*.{h,m}"
  s.docset_url   = "https://geofire-ios.firebaseapp.com/docs/"
  s.ios.deployment_target = '7.0'
  s.ios.dependency  'Firebase', '~> 3.2'
  s.framework = 'CoreLocation'
  s.requires_arc = true
end
```

**Comment by mcdonamp** [comment]
@HeyDaniel thanks for doing this, lots of folks (see #37) have wanted it, and we're glad that you put forth the effort to 🚢🔥

A few bits of cleanup as @jdimond and I have mentioned, but we should be ready to go soon!

**Comment by TimKreger-IA** [comment]
Hi all,
just a question about updating the tests. At the moment the Real Data tests wont work with the 3.2.0 as they require FireApp configuration to work with the upgraded database. There doesn't seem to be a simple way around this for an open source project unless there is a database project created for the repo.  I'm guessing this could be difficult or risky now that the API now resides fully within the Googlesphere. Anyone have any ideas as to how to do integration tests on this type of project?

File: GeoFire/GFCircleQuery.h (hunk 1 above)
Comment (HeyDaniel) [inline GeoFire/GFCircleQuery.h]: cause i don't know how to run my code with the private header 😅.

File: GeoFire/GeoFire.h (hunk 2 above)
Comment (HeyDaniel) [inline GeoFire/GeoFire.h]: seem like you don't implement + (NSDictionary *)dictFromLocation:(CLLocation *)location

**Comment by HeyDaniel** [comment]
i have signed the google CLA.

**Comment by HeyDaniel** [comment]
instantly reply me,this google bot is amazing huh?

**Comment by greggmojica** [comment]
@HeyDaniel Thanks for doing this, and great job, Team Firebase!!

**Comment by HeyDaniel** [comment]
@greggmojica  Thank You for building Firebase ,a Great backend cloud service .

**Comment by mcdonamp** [comment]
Ah yeah, tests... So the short answer is that one can programmatically create an app that can be used (using Swift since I've already got an example for it):

```
let options = FIROptions(googleAppID:  bundleID: , GCMSenderID: , APIKey: , clientID: , trackingID: , androidClientID: , databaseURL: "https://testNamespace.firebaseio.com", storageBucket: , deepLinkURLScheme: )   // fill in all the other fields...
FIRApp.configureWithName("testApp", options: options)
let app = FIRApp(named: "testApp")
FIRDatabase.database(app: app!)   // gets you a "programmatically" created database
```

We should update the tests to ensure that they run. We can futz with CI to always pick up the correct config when we run tests (or just always run it locally and replace with a snippet we/the person developing tests with).

**Comment by TimKreger-IA** [comment]
@mcdonamp Thanks Michael, I looked at FIROptions and the comment on the said the app, bundle and GCM sender are required but didn't try it. Will give it a go, I was also thinking that an assert requiring the googleserves for the plist might work but a required define and programmatic config might be the way to go.

**Comment by cxt9** [comment]
Any update on publishing the official version? 

How can I embed the current one into an existing project? Adding the files to my project and using a bridge file still failed.

**Comment by greggmojica** [comment]
@mcdonamp Same here.

**Comment by mcdonamp** [comment]
There are still unresolved issues before we can merge (and by extension deploy):
- Move files back to `/Implementation`
- Revert changes to `+Private.h` extensions
- Add `Podspec` changes
- Convert unit tests to run on a Firebase 3.x client

@HeyDaniel, if you're not willing/able to do these (and that's fine, we're all busy people), let's close this PR so that someone else can do it; or if someone else implements these things in the mean time, I'm happy to look at those changes. I think we've made our expectations clear :)

On a somewhat related note: this is open source software--anyone (yourself included!) can clone the repo, modify the code, install the pod locally, ship to production, submit a PR, and contribute back. As I said before, this is a great way to gain more skills and contribute back to the community.

**Comment by TimKreger-IA** [comment]
Hi all, I have a fork which is updated and running with the original structure, with tests still outstanding. If @HeyDaniel  doesn't have time to do this I can finish off the tests and do a pull request.

**Comment by mcdonamp** [comment]
@TimKreger-IA I would just submit a PR with the updates (including tests) and I can do it in parallel.

**Comment by HeyDaniel** [comment]
@mcdonamp i actually don't know how to use private.h/add Podspec and write unit test to this project,close the PR is fine.

**Pull request closed without merging**